### ISSUE TITLE
BUG: Failed "pip install" build from source

### ISSUE THREAD

**Issue by thomasnal** [scipy/scipy#18005]
### Describe your issue.

I want to pip install scipy. Pip is building the wheel from source. scipy build process fails.
```sh
$ OPENBLAS=$(brew --prefix openblas) pip install scipy==1.7.3
Collecting scipy==1.7.3
  Using cached scipy-1.7.3.tar.gz (36.1 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy<1.23.0,>=1.16.5 in /Users/thomas/.pyenv/versions/3.7.16/envs/trading-system-3.7.16/lib/python3.7/site-packages (from scipy==1.7.3) (1.21.6)
Building wheels for collected packages: scipy
  Building wheel for scipy (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for scipy (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [10718 lines of output]
...
      /opt/homebrew/bin/gfortran -Wall -g -L/opt/homebrew/opt/zlib/lib build/temp.macosx-12.3-arm64-3.7/scipy/integrate/_quadpackmodule.o -L/opt/homebrew/opt/openblas/lib -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12 -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12/../../.. -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12/../../.. -L/Users/thomas/.pyenv/versions/3.7.16/lib -Lbuild/temp.macosx-12.3-arm64-3.7 -Wl,-rpath,/opt/homebrew/opt/openblas/lib -lquadpack -lmach -lopenblas -lopenblas -lgfortran -o build/lib.macosx-12.3-arm64-3.7/scipy/integrate/_quadpack.cpython-37m-darwin.so
      Undefined symbols for architecture arm64:
        "_PyArg_ParseTuple", referenced from:
            _quadpack_qagse in _quadpackmodule.o
...
        "_main", referenced from:
           implicit entry/start for main executable
      ld: symbol(s) not found for architecture arm64
...
      error: Command "/opt/homebrew/bin/gfortran -Wall -g -L/opt/homebrew/opt/zlib/lib build/temp.macosx-12.3-arm64-3.7/scipy/integrate/_quadpackmodule.o -L/opt/homebrew/opt/openblas/lib -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12 -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12/../../.. -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12/../../.. -L/Users/thomas/.pyenv/versions/3.7.16/lib -Lbuild/temp.macosx-12.3-arm64-3.7 -Wl,-rpath,/opt/homebrew/opt/openblas/lib -lquadpack -lmach -lopenblas -lopenblas -lgfortran -o build/lib.macosx-12.3-arm64-3.7/scipy/integrate/_quadpack.cpython-37m-darwin.so" failed with exit status 1
```
Looks like the linker is missing `-lpython` to build a lib instead of an executable?

### Reproducing Code Example

```python
$ OPENBLAS=$(brew --prefix openblas) pip install scipy==1.7.3
```

### Error message

```shell
Collecting scipy==1.7.3
  Using cached scipy-1.7.3.tar.gz (36.1 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy<1.23.0,>=1.16.5 in /Users/thomas/.pyenv/versions/3.7.16/envs/trading-system-3.7.16/lib/python3.7/site-packages (from scipy==1.7.3) (1.21.6)
Building wheels for collected packages: scipy
  Building wheel for scipy (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for scipy (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [10718 lines of output]
...
      /opt/homebrew/bin/gfortran -Wall -g -L/opt/homebrew/opt/zlib/lib build/temp.macosx-12.3-arm64-3.7/scipy/integrate/_quadpackmodule.o -L/opt/homebrew/opt/openblas/lib -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12 -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12/../../.. -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12/../../.. -L/Users/thomas/.pyenv/versions/3.7.16/lib -Lbuild/temp.macosx-12.3-arm64-3.7 -Wl,-rpath,/opt/homebrew/opt/openblas/lib -lquadpack -lmach -lopenblas -lopenblas -lgfortran -o build/lib.macosx-12.3-arm64-3.7/scipy/integrate/_quadpack.cpython-37m-darwin.so
      Undefined symbols for architecture arm64:
        "_PyArg_ParseTuple", referenced from:
            _quadpack_qagse in _quadpackmodule.o
...
        "_main", referenced from:
           implicit entry/start for main executable
      ld: symbol(s) not found for architecture arm64
...
      error: Command "/opt/homebrew/bin/gfortran -Wall -g -L/opt/homebrew/opt/zlib/lib build/temp.macosx-12.3-arm64-3.7/scipy/integrate/_quadpackmodule.o -L/opt/homebrew/opt/openblas/lib -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12 -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12/../../.. -L/opt/homebrew/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/aarch64-apple-darwin21/12/../../.. -L/Users/thomas/.pyenv/versions/3.7.16/lib -Lbuild/temp.macosx-12.3-arm64-3.7 -Wl,-rpath,/opt/homebrew/opt/openblas/lib -lquadpack -lmach -lopenblas -lopenblas -lgfortran -o build/lib.macosx-12.3-arm64-3.7/scipy/integrate/_quadpack.cpython-37m-darwin.so" failed with exit status 1
```

### SciPy/NumPy/Python version information

n/a
1.21.6
3.7.16 (default, Feb 19 2023, 17:20:38)
[Clang 13.1.6 (clang-1316.0.21.2.5)]

**Comment by rgommers** [comment]
I'll note that there's a wheels for macOS >=12.0 at https://pypi.org/project/scipy/1.7.3/#files. macOS 11 is no good, we had to avoid that due to a severe OS bug - please upgrade to macOS 12 if you want to use SciPy.

SciPy 1.7.3 is also too old, and the `setup.py` build is no longer supported. I don't recognize this build error, but I'll just say that it's not worth investigating anymore.

**Comment by thomasnal** [comment]
I'm on macOS 12.3
<img width="212" alt="image" src="https://user-images.githubusercontent.com/17355467/219974380-5a87df6e-8f4a-4e93-b4d9-8b03f6bd9377.png">

**Comment by thomasnal** [comment]
Also it is not me who picks this scipy version but pip.  I want sklearn module. I do `pip install scikit-learn` and it selects scipy 1.7.3.  So then I isolated to `pip install scipy==1.7.3`.

**Comment by andyfaff** [comment]
You're going to have to upgrade your Python interpreter. There are `cp38-macosx_12_0_arm64` wheels (and upwards) readily available from scipy-1.7.3 onwards, but none for Python 3.7. No `macosx_arm64` wheels were available before 1.7.3.

Homebrew has a more recent Python interpreter, or you could use a Python.org supplied install. Personally I use `conda/miniconda`.

**Comment by thomasnal** [comment]
For the project and its dependencies I need to use py37.

**Comment by rgommers** [comment]
> There are `cp38-macosx_12_0_arm64` wheels (and upwards) readily available from scipy-1.7.3 onwards, but none for Python 3.7.

Right, good correction @andyfaff, I misread the wheel tags. Indeed, we never had a working Python 3.7 build for macOS arm64.

> For the project and its dependencies I need to use py37.

You can't reasonably do this, Python 3.7 really is too old to run on macOS arm64. Nothing works; even getting 3.8 to work was tricky.
